src/readers/morphologyHDF5.cpp

```diff
@@ -275,126 +274,109 @@ void MorphologyHDF5::_readPoints(int firstSectionOffset) {
-            } catch (HighFive::DataSetException&) {
-                if (_stage == "unraveled") {
-                    std::string raw_path = "/" + _g_root + "/" + _g_structure + "/raw";
-                    try {
-                        return _group.getDataSet(raw_path);
-                    } catch (HighFive::DataSetException&) {
-                        throw(MorphioError("Could not find unraveled structure neither at " + path +
-                                           " or " + raw_path + " for dataset for morphology '" +
-                                           _uri + "' repair stage " + _stage));
-                    }
-                } else {
-                    throw(MorphioError("Could not open " + path + " dataset for morphology '" +
-                                       _uri + "' repair stage " + _stage));
-                }
-            }
-        }();
-
-        _sections.reset(new HighFive::DataSet(dataset));
-
-        const auto dims = dataset.getSpace().getDimensions();
-
-        if (dims.size() != 2 || dims[1] != _structureV2Columns) {
-            throw(MorphioError("Error reading morphologies " + _uri +
-                               " bad number of dimensions in 'structure' dataspace"));
-        }
-
-        std::vector<std::vector<int>> vec;
-        vec.resize(dims[0] - 1);
-        dataset.read(vec);
-        int firstSectionOffset = vec[1][0];
-        sections.reserve(sections.size() + vec.size() - 1);
-        bool skipFirst = true;
-        for (const auto& p : vec) {
-            if (skipFirst) {
-                skipFirst = false;
-                continue;
-            }
-            sections.emplace_back(Property::Section::Type{p[0] - firstSectionOffset, p[1] - 1});
-        }
-
-        return firstSectionOffset;
-    }
+inline int MorphologyHDF5::_readSections() {
+    return (_properties.version() == MORPHOLOGY_VERSION_H5_2) ? _readV2Sections()
+                                                              : _readV1Sections();
+}
 
-    auto selection = _sections->select({0, 0}, {_sectionsDims[0], 2}, {1, 2});
+int MorphologyHDF5::_readV1Sections() {
+    auto& sections = _properties.get<Property::Section>();
+    auto& types = _properties.get<Property::SectionType>();
 
     std::vector<std::vector<int>> vec;
-    vec.resize(_sectionsDims[0]);
-    selection.read(vec);
+    _sections->read(vec); 
 
     if (vec.size() < 2)  // Neuron without any neurites
         return -1;
 
     int firstSectionOffset = vec[1][0];
-
     sections.reserve(sections.size() + vec.size() - 1);
-    bool skipFirst = true;  // Skipping soma section
-    for (const auto& p : vec) {
-        if (skipFirst) {
-            skipFirst = false;
-            continue;
+    types.reserve(vec.size() - 1);  // remove soma type
+
+    for (size_t i = 1; i < vec.size(); ++i) {
+        const auto& p = vec[i];
+        const auto& type = p[1];
+
+        if (type > SECTION_CUSTOM_START || type < 0) {
```

Comment:
From clang:
```
/home/travis/build/BlueBrain/MorphIO/src/readers/morphologyHDF5.cpp:368:49: error: 
      result of comparison of unsigned enum expression < 0 is always false
      [-Werror,-Wtautological-unsigned-enum-zero-compare]
        if (type > SECTION_CUSTOM_START || type < 0) {
                                           ~~~~ ^ ~
```
Prev. types was getting cast to int, which seems like a bug: https://github.com/BlueBrain/MorphIO/pull/194/files#diff-fecd1a78269c91432367e6141e692f06L378

I think @wizmer will have to look at that.

Reply:
This is strange, because at the end this is pointing to a traditional `enum` and the values can be negative numbers, right? Even replacing the `0` with `SECTION_UNDEFINED` (which is 0) might not work if the compiler is assuming unsigned values, strange!

Reply:
`traditional enum` are implementation defined, it seems; so in this case clang chose `unsigned` it seems.

I think the logic of this needs to be checked; paging @wizmer.

Reply:
If the files are expected to contain negative values for the types, I believe there is a way to enforce "typed enums", so that you can tell the compiler that it is a signed integer. But this is a C++11 feature I think.

Reply:
I don't manage to reproduce the clang error message on my machine.
I also don't understand how clang comes to the conclusion that `type` is an enum given that:
```cpp
        std::vector<std::vector<int>> vec;
        const auto& p = vec[i];
        const auto& type = p[1];
```
So `type` should be an `int` in the first place. Actually @sergiorg-hpc , could you replace the two `auto` by their actual type, that will make things more explicit ? Maybe it's the comparison to `SECTION_CUSTOM_START` that is messing up with clang ?

But yes, I think the code is correct. The idea is to fill `types` with enum elements and to compare the integer value of each enum to see if they are out of range. I have added a test for negative values in this PR: https://github.com/BlueBrain/MorphIO/pull/195

Reply:
I think the comparison is the actual problem. To give it a try, I just added a cast to `int` for the `SECTION_CUSTOM_START`, making the comparison to plain integers instead of `enum` types. This will maintain the original logic of the source code, without additional changes.

Reply:
> I don't manage to reproduce the clang error message on my machine.

It's in the log (clang 7)
https://travis-ci.com/github/BlueBrain/MorphIO/jobs/379841972#L446

> I also don't understand how clang comes to the conclusion that type is an enum given that:
> std::vector<std::vector<int>> vec;

Shouldn't the type of that vector be something more specific than an `int`?

Reply:
Perhaps @matz-e can elaborate this, but as far as I understand it, the `structure` dataset inside the morphology files contain integer values. This is just to read the file, then it is converted to the specific datatype.

However, the trick of casting to `int` didn't work. The compiler is assuming that `0` is somehow unsigned.

Reply:
@wizmer's suggestion of forcing the type to `int` seemed to work, thanks a lot! Closing down this thread.